src/vm/crossloaderallocatorhash.inl

```diff
@@ -0,0 +1,1216 @@
+
+    bool result = true;
+
+    GCPROTECT_BEGIN(gc);
+    {
+        if (gc.hashKeyEntry->GetMethodTable() == MscorlibBinder::GetExistingClass(CLASS__LAHASHKEYTOTRACKERS))
+        {
+            gc.hashKeyToTrackers = (LAHASHKEYTOTRACKERSREF)gc.hashKeyEntry;
+            gc.keyValueStore = gc.hashKeyToTrackers->_laLocalKeyValueStore;
+        }
+        else
+        {
+            gc.keyValueStore = gc.hashKeyEntry;
+        }
+
+        result = VisitKeyValueStore(&gc.loaderAllocatorRef, &gc.keyValueStore, visitor);
+    }
+    GCPROTECT_END();
+
+    return result;
+}
+
+template <class TRAITS>
+/*static*/ void CrossLoaderAllocatorHash<TRAITS>::DeleteEntryTracker(TKey key, LAHASHDEPENDENTHASHTRACKERREF trackerUnsafe)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_COOPERATIVE;
+    }
+    CONTRACTL_END;
+
+    struct 
+    {
+        LAHASHDEPENDENTHASHTRACKERREF tracker;
+        OBJECTREF loaderAllocatorRef;
+        GCHEAPHASHOBJECTREF keyToValuesHashObject;
+        KeyToValuesGCHeapHash keyToValuesHash;
+    }gc;
+
+    ZeroMemory(&gc, sizeof(gc));
+    gc.tracker = trackerUnsafe;
+
+    GCPROTECT_BEGIN(gc);
+    {
+        gc.tracker->GetDependentAndLoaderAllocator(&gc.loaderAllocatorRef, &gc.keyToValuesHashObject);
+        if (gc.keyToValuesHashObject != NULL)
+        {
+            gc.keyToValuesHash = KeyToValuesGCHeapHash(gc.keyToValuesHashObject);
+            gc.keyToValuesHash.DeleteEntry(&key);
+        }
+    }
+    GCPROTECT_END();
+}
+
+#ifndef DACCESS_COMPILE
+/*static */inline void CrossLoaderAllocatorHashSetup::EnsureTypesLoaded()
+{
+    STANDARD_VM_CONTRACT;
+
+    // Force these types to be loaded, so that the hashtable logic can use MscorlibBinder::GetExistingClass
+    // throughout and avoid lock ordering issues
+    MscorlibBinder::GetClass(CLASS__LAHASHKEYTOTRACKERS);
+    MscorlibBinder::GetClass(CLASS__LAHASHDEPENDENTHASHTRACKER);
+    MscorlibBinder::GetClass(CLASS__GCHEAPHASH);
+    TypeHandle elemType = TypeHandle(MscorlibBinder::GetElementType(ELEMENT_TYPE_I1));
+    TypeHandle typHnd = ClassLoader::LoadArrayTypeThrowing(elemType, ELEMENT_TYPE_SZARRAY, 0);
+    elemType = TypeHandle(MscorlibBinder::GetElementType(ELEMENT_TYPE_OBJECT));
+    typHnd = ClassLoader::LoadArrayTypeThrowing(elemType, ELEMENT_TYPE_SZARRAY, 0);
+}
+
+template <class TRAITS>
+void CrossLoaderAllocatorHash<TRAITS>::EnsureManagedObjectsInitted()
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_COOPERATIVE;
+    }
+    CONTRACTL_END;
+
+    if (m_loaderAllocatorToDependentTrackerHash == NULL)
+    {
+        OBJECTREF laToDependentHandleHashObject = AllocateObject(MscorlibBinder::GetExistingClass(CLASS__GCHEAPHASH));
+        m_loaderAllocatorToDependentTrackerHash = m_pLoaderAllocator->GetDomain()->CreateHandle(laToDependentHandleHashObject);
+        m_pLoaderAllocator->RegisterHandleForCleanup(m_loaderAllocatorToDependentTrackerHash);
```

Comment:
I see only one call to `AssemblyLoaderAllocator::CleanupHandles` from the `AppDomain` and from the name it seems to be relevant to shutdown. How would these hash tables be freed for collectible `LoaderAllocator`s?

Reply:
There are 2 types of GCHandles available for use from within the LoaderAllocator logic. There are LoaderAllocator Handles and GC Handles. LoaderAllocator Handles have lifetime that will last till the end of the managed lifetime of the loaderallocator, but their finalizer may be triggered earlier, as part of general LoaderAllocator finalization. (Its even possible through finalization resurrection to make various things come back to semi-life. However, in this path, we use a GC Handle which is registered for cleanup AFTER the LoaderAllocator is considered destroyed by the managed environment. These handles are cleaned up during a later phase of LoaderAllocator shutdown which happens to be managed by the AppDomain class. (The restriction on these things is that they MUST NOT have any managed pointers to managed objects that can keep a LoaderAllocator alive.) Thus yes, the logic in AppDomain.cpp that calls CleanupHandles is relevant to shutdown, but that shutdown refers to LoaderAllocator shutdown, not AppDomain shutdown.